### PR TITLE
fix(agent-cli): a subagent runs on the provider the parent composed (ARCH-109)

### PR DESCRIPTION
**Pull request by woojubb**
The parent resolved a provider composition its children never saw, and the children rebuilt a different one without saying so.

## What crosses, and what did not

| Where | What it does |
| ----- | ------------ |
| the shell entry point | resolves `providerDefinitions`; the product may additionally receive `provider: loadReplayProvider(args.sessionLog)` |
| the child runner factory | receives `IProviderDefinitionConfig` — settings, including `apiKey` — never the definition set and never the replay provider |
| `createRobotaSubagentComposition()` | returned `createDefaultProviderDefinitions()` unconditionally |
| the worker entry | called it **with no arguments** |

Two consequences, both reachable at the revision this was measured on:

**A `--session-log` replay parent spawned children that called the live provider with a real key.** The child runner is built from the settings-derived config *before* the replay override is folded into the product, so the parent replayed deterministically while its subagents made billed calls. The comment three lines above that override said "no key is ever used" — true of the parent, false of its children.

**A `startCli({ providerDefinitions })` caller got a parent that resolved and a child that did not** — or worse than an error, a child resolving a *different* implementation that happened to share a provider name.

## Why the recipe is not serialized

`IProviderDefinition` carries `createProvider` and `probeProfile`, both **function-valued**. "Send the parent's definitions to the child" is unimplementable, and saying so first is what rules out the shape the issue proposed.

The seam that does work already existed **and already named this case**. `ISubagentWorkerComposition` documents `providerDefinitions` as carrying definitions rather than a constructed provider precisely so *"a custom provider type resolves instead of throwing `Unknown provider`"*, and the capability guard's own docblock calls `sandboxFactories` *"the same shape, and the same reason, as `providerDefinitions`"*. Robota's own worker entry pinned it to the default set, so the seam was present and unused. It is a parameter now, for the reason the pack factory beside it already is.

## The correction the evidence forced

**The first implementation refused at composition time**, matching the capability guard next to it. Twenty existing tests drive the CLI through `startCli({ providerDefinitions })` — the supported way to embed the product with your own providers — and **none of them spawn a subagent**. The throw was a false positive against working software, and the count made that unarguable rather than a matter of taste.

A narrower version warned instead. Four more tests failed: they assert an **empty stderr** for print and JSON runs, where stderr is part of the output contract.

Both measurements say the same thing — a caller-supplied set is the normal path, not an anomaly. The defect was never that a caller composed providers; it is that the **child** resolves a different one.

So it **selects rather than refuses**: a composition a child cannot rebuild runs its subagents in-process, sharing the parent's provider. The defect is removed rather than reported, and every legitimate caller keeps working. What is given up is process isolation for subagents, so that is said out loud rather than inferred.

Three details that are load-bearing:

- `buildChildProcess` is a **thunk**, so the fallback branch never constructs the child runner — and constructing it is what reads `apiKey`. "No live config was assembled" is a stronger property than "the returned runner is the in-process one", and it is the one the test holds.
- The notice fires for a composition the **operator** caused (`--session-log`, typed by a person, changing what they observe) and not for one an embedding **program** made, which it cannot act on today and which would break the print/JSON stderr contract.
- **The capability guard still throws.** The asymmetry is deliberate: an unreproducible sandbox is a containment property where the safe direction is to stop; an unreproducible provider has a correct fallback, and stopping would be the unsafe direction for the user's session.

## Verification

`pnpm harness:verify-like-ci` — **PASS, all 13 stages**. Package suite 435 passed across 56 files. Repo lint 2201 warnings against the 2203 ceiling.

**Falsification, three arms, each verified to have applied before its verdict was read** (a mutation whose commit or write silently fails measures the treatment twice — that happened earlier in this session and is why the mutation site is asserted first):

| Arm | Mutation | Result |
| --- | -------- | ------ |
| A | the selector ignores reproduction and always builds the child runner | 2 failed / 3 passed |
| B | the predicate never reports a replay provider | 2 failed / 3 passed |
| C | the child recipe pins the default set again | 1 failed / 4 passed |
| — | treatment | 5 passed |

Two assertions are deliberately stronger than the obvious form: the fallback is asserted as a **zero call count** on the child-runner builder, and the recipe's definitions are asserted by **identity** — a name match passes for the default set too and so could not fail on the defect it names.

## Size floor

The shell entry point is frozen as debt, where the rule is split rather than extend. The runner wiring moved out of it — the log directory, self-fork entry and worktree adapter are robota's own answers with no reason to be spelled at the call site — and the reproduction question moved to its own module. The file lands **back at its baseline**, not above it. `createRobotaChildProcessSubagentRunner` is no longer exported: it is one branch of a selection, and leaving it exported would let a call site take that branch with no selection in front of it.

Refs issue #2044

### CODE REVIEW

File: packages/agent-cli/src/product/__tests__/subagent-provider-reproduction.test.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+   *
+   * The shared name is the point. It is the case a name-comparison check would call "same" and the
+   * child would then run different code for — which is why the ORIGIN is reported by the composition
+   * root rather than re-derived from the definitions downstream.
+   */
+  const CALLER_DEFINITIONS = [
+    { type: 'openai', createProvider: () => ({}) },
+  ] as unknown as readonly IProviderDefinition[];
+
+  function record(): { readonly notices: string[]; readonly notice: (m: string) => void } {
```
[the remark this answers is not in the document]
Comment (woojubb) [reply packages/agent-cli/src/product/__tests__/subagent-provider-reproduction.test.ts]: Correct, and fixed in `df3f8ecfe`.

`record` was declared in both describe blocks when the ARCH-109 tests moved into this file. The first block never uses it — its three cases assert the predicate’s return value directly and take no notice callback — so only the second declaration was ever reached. Removed the first; the file keeps one.

Worth noting which reviewer caught it. The human-facing review read the branching and the properties and returned `ACTIONABLE FINDINGS: 0`, which was right about the question it was asked: an unused declaration is not a property of the change’s behaviour. This is what the static analyser is for, and the two are not redundant.

5 tests still pass; eslint clean on the file.
